### DISCUSSION TITLE
React-Zustand able to mutate state without set()

### DISCUSSION THREAD

**Discussion by ishankgupta95** [pmndrs/zustand#2611] [Q&A] [answered]
I can update the state without using set(). Is this allowed? I was under the impression that the set() must be used to update the store since the state is immutable

I have a code [sandbox](https://codesandbox.io/p/devbox/xf9k3n?file=%2Fpackage.json&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxm9pujk00063b6m4o35sr9x%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxm9pujj00023b6m5c7odia8%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxm9pujj00043b6mmj3mqz78%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxm9pujj00053b6m149tlo8d%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B63.24566131292444%252C36.75433868707556%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxm9pujj00023b6m5c7odia8%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxm9pujj00013b6mbp1x6whg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clxm9pujj00023b6m5c7odia8%2522%252C%2522activeTabId%2522%253A%2522clxm9pujj00013b6mbp1x6whg%2522%257D%252C%2522clxm9pujj00053b6m149tlo8d%2522%253A%257B%2522id%2522%253A%2522clxm9pujj00053b6m149tlo8d%2522%252C%2522activeTabId%2522%253A%2522clxm9v70f007o3b6m3qrvh2ls%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522port%2522%253A5173%252C%2522taskId%2522%253A%2522Development%2522%252C%2522id%2522%253A%2522clxm9v70f007o3b6m3qrvh2ls%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clxm9pujj00043b6mmj3mqz78%2522%253A%257B%2522id%2522%253A%2522clxm9pujj00043b6mmj3mqz78%2522%252C%2522activeTabId%2522%253A%2522clxm9q5jp005f3b6mbpl50o5x%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxm9pujj00033b6mmrt8agi7%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clxm9puwx000adif8cba02vmu%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522Development%2522%252C%2522id%2522%253A%2522clxm9q5jp005f3b6mbpl50o5x%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) here which isolates the issue I am talking about

I have three components and an array in store, I am sorting the array from the store in one of the three components and the state in other components is also being sorted which means that the store is updating without set()?

I might be fundamentally wrong here about the state updation

**Comment by dai-shi** [comment, marked as answer]
Yes, our contract (and React's one) is to treat state as immutable object.
So, you shouldn't mutate state object. You can only replace it with `set()`.
Mutating state can go wrong and break some behaviors.

In your specific example, you should use `.toSorted()` or `[...arr].sort()`.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toSorted

**Comment by r00ott** [comment]
Yes, both our contract and React's recommend treating state as an immutable object. Therefore, you should avoid mutating the state directly. Instead, use the `set()` method to replace it. Mutating the state directly can cause unexpected issues and disrupt functionality.

For your particular example, you should use `.toSorted()` or `[...arr].sort()`. You can find more information by searching about : "MDN Web Docs on Array.toSorted"
